### ISSUE TITLE
Special keys do not work anymore

### ISSUE THREAD

**Issue by trudnorx** [mumble-voip/mumble#5020]
Since the input system got updated, I can still use add hotkeys with regular keys, but special keys from the keyboard cannot be added anymore.

I'm talking here about special, non-standard, custom keyboard keys that only come with some keyboards, assignable to various user functions. In the old Mumble they used to get detected as "Unknown - #xxx" or something like that, with Mumble not knowing what they are, with a special number, but they used to work.

Steps to reproduce:
1. Try to add a Mumble hotkey with a special keyboard key.
2. Notice the key is not detected at all and Shortcut stays at "No buttons assigned".

**Comment by trudnorx** [comment]
And the update deletes old hotkeys.

**Comment by davidebeatrici** [comment]
Old hotkeys are deleted because they are now saved differently (it depends on the backend), however all keys are supposed to appear because we're using raw input.

What keyboard is it?

**Comment by trudnorx** [comment]
I don't think it depends on keyboard model because this keyboard doesn't have any custom driver or anything. So probably there are just some extra keys which are sent with keycodes that are on the 'standard' table but do not have any standard assignment (or something like that).

And here's another issue: [AppKeys](https://www.google.com/search?q=appkey+windows&hl=en) like e.g. search are failing too -- getting detected as 'K0'.

So while some special keys are not detected at all, there are multiple special keys which are all falsely detected as K0.

**Comment by trudnorx** [comment]
Volume up/down keys are falsely getting detected as K0. You can test that on almost any keyboard. Media play/pause key is not detected at all. Again, testable on almost any keyboard.

**Comment by davidebeatrici** [comment]
App keys can be enabled by specifying `RIDEV_APPKEYS` when registering the raw input device: https://docs.microsoft.com/en-us/windows/win32/api/winuser/ns-winuser-rawinputdevice

The problem is that `RIDEV_NOLEGACY` has to be specified too and that causes input not to be registered by Qt anymore.

As a workaround we can handle `WM_APPCOMMAND` messages: https://docs.microsoft.com/en-us/windows/win32/inputdev/wm-appcommand

**Comment by trudnorx** [comment]
Actually it looks like media play/pause is detected as K0 too so it's not the same issue as the "special keys" that are not detected at all. Interestingly, however, media play/pause wasn't getting detected at all as long as another app was using it. Maybe the issues are linked. Maybe some app is using my special keys? Did Mumble go from being able to detect keys which are in use by other apps to not at all? Or maybe media play/pause is somehow 'exclusive' in a way other keys aren't.

**Comment by trudnorx** [comment]
Testing reveals media play/pause key cannot be caught by WM_APPCOMMAND if it is in use by another application. Also, it cannot detect any of the 'special keys' I talked about at all. But this may be regular behavior because these may not be AppKeys.

Is your raw input code able to detect keycodes >255?

**Comment by davidebeatrici** [comment]
Yes, it's a 16 bit variable.

**Comment by trudnorx** [comment]
https://www.nirsoft.net/utils/keyboard_state_view.html is able to detect appkeys but none of the special keys. So fixing the appkeys issue won't necessarily fix the special keys, I guess.

**Comment by trudnorx** [comment]
Looks like the keyboard registers as RIM_TYPEHID and sends the special keys, & other raw input programs are able to catch it, but for whatever reason the new Mumble isn't catching them.

**Comment by davidebeatrici** [comment]
The `K` prefix means that the packet is received as `RIM_TYPEKEYBOARD`.

With my keyboard multimedia keys appear as follows:

- Stop: `J`
- Reverse: `Q`
- Play/Pause: `G`
- Forward: `P`
- Mail: `108`
- Home: `M`
- Calculator: `F`
- Volume-: `C`
- Volume+: `B`
- Mute:  `D`

**Comment by trudnorx** [comment]
> The `K` prefix means that the packet is received as `RIM_TYPEKEYBOARD`.
Ok but that's for AppKeys which are falsely detected as K0. What I defined as "special keys" (which excludes AppKeys, by my definition) is not detected at all, thus no K. And, after reading on the web, it seems that some keyboards establish a RIM_TYPEHID device, in addition to RIM_TYPEKEYBOARD, and send the 'special keys' through RIM_TYPEHID. I think that's what's happening here, because other raw input apps are detecting a RIM_TYPEHID *and* are detecting the special key presses there, yet new Mumble is not. (but old Mumble, before the implementation of raw input, was, as I explained)

Do you understand? And do you have any ideas what could be causing that?

**Comment by davidebeatrici** [comment]
Keyboards that define extra keys should indeed have an extra interface that is detected as `RIM_TYPEHID`.

What I don't understand is why in your case they appear as `RIM_TYPEKEYBOARD`.

Which raw input apps are working fine?

**Comment by trudnorx** [comment]
> Keyboards that define extra keys should indeed have an extra interface that is detected as `RIM_TYPEHID`.
> What I don't understand is why in your case they appear as `RIM_TYPEKEYBOARD`.

I believe that you are not reading what I stated carefully enough. In my case, they **are* RIM_TYPEHID, and Mumble is not detecting them **at all** (neither as RIM_TYPEKEYBOARD nor as anything else). This is a _different_ issue to my media keys being detected as RIM_TYPEKEYBOARD (but with the wrong keycode).

**Comment by trudnorx** [comment]
Let me attempt to clarify the discussion by stating the two issues clearly

**Issue 1**. -- My keyboard's extra keys, which are RIM_TYPEHID, show up in other raw input apps; but Mumble is completely unable to detect them; they show up as **No buttons assigned** (old Mumble, before raw input, was able to use them as shortcut).

**Issue 2**. -- AppKeys, e.g. volume up/down, are successfully detected by Mumble, but falsely all detected as "K0". They show up as "**K0**".

**Comment by trudnorx** [comment]
Do you want me to split these two issues by creating two different new threads?

**Comment by davidebeatrici** [comment]
No, it's fine.

Is the HID device detected in the developer console?

**Comment by trudnorx** [comment]
I did a DebugView (is that ok?) and I see nothing about HID or keyboards.

**Comment by davidebeatrici** [comment]
I would use Mumble's console, just to be sure.

**Comment by trudnorx** [comment]
How do you enable it? I did DebugView because that's what I found when I googled "mumble console".

**Comment by davidebeatrici** [comment]
![](https://user-images.githubusercontent.com/5897523/119214036-ec40f080-bac3-11eb-92f2-8dfd6eb04c1a.png)

Check "Enable Developer menu".

![](https://user-images.githubusercontent.com/5897523/119214069-2f02c880-bac4-11eb-9abe-e4f943baffc6.png)

**Comment by trudnorx** [comment]
I don't see anything about any keyboard or HID in the console.
